doc/guides/naming_conventions/index.rst

```diff
@@ -0,0 +1,88 @@
+Naming Conventions
+##################
+
+Functions and Procedures
+************************
+
+#. No procedure shall have a name that is a keyword of any standard version of
+   the C or C++ programming language. Restricted names include ``interrupt``, ``inline``,
+   ``class``, ``true``, ``false``, ``public``, ``private``, ``friend``,
+   ``protected``, and many others.
+
+#. Function names are always in lowercase, with an underscore between each part of
+   the name : ``k_timer_init()`` , ``k_timer_start()``.
+
+#. If a function operates primarily on a certain type, the prefix of this function
+   corresponds to the type name. For example, the k_timer_* functions work with the
+   ``k_timer`` type, and k_sem_* functions go with ``k_sem``.
+
+#. Each procedure’s name shall be descriptive of its purpose. Note that procedures
+   encapsulate the “actions” of a program and thus benefit from the use of verbs in
+   their names (e.g., ``k_mutex_lock()``); this “noun-verb” word ordering is recommended.
+   Alternatively, procedures may be named according to the question they answer
+   (e.g., ``device_is_ready()``).
+
+#. The names of all public functions shall be prefixed with their module name and
+   an underscore (e.g., ``sensor_read()``).
+
+#. Avoid names that differ only in case, like foo and Foo. Similarly, avoid foobar
+   and foo_bar. The potential for confusion is considerable.
+
+#. Similarly, avoid names that look like each other. On many terminals and
+   printers, 'l', '1' and 'I' look quite similar. A variable named 'l' is
+   particularly bad because it looks so much like the constant '1'.
+
+#. #define constants should be in all CAPS.
```

Comment:
I'd really like an exception for this when the constant provides the numeric value of a measurement expressed in specific units, so we can use the mixed case unit specifiers from [the relevant standard](https://ucum.org/ucum.html).  E.g. I really want `MAX_TEMP_cCel` or `SAMPLE_RATE_kHz`.

This also applies to variable names, function names, anything else where a unit is encoded.

Reply:
Well, it's a "should" so I think your case is covered. Aside from this, I agree.

Reply:
I've had changes requested when I've done this in the past, so I'd prefer something explicit.